swap/src/network/tor_transport.rs

```diff
@@ -0,0 +1,225 @@
+use data_encoding::BASE32;
+use futures::future::Ready;
+use futures::prelude::*;
+use libp2p::core::multiaddr::{Multiaddr, Protocol};
+use libp2p::core::transport::TransportError;
+use libp2p::core::Transport;
+use libp2p::tcp::tokio::{Tcp, TcpStream};
+use libp2p::tcp::{GenTcpConfig, TcpListenStream, TokioTcpConfig};
+use std::cmp::Ordering;
+use std::io;
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+use std::pin::Pin;
+use tokio_socks::tcp::Socks5Stream;
+use tokio_socks::IntoTargetAddr;
+
+/// Represents the configuration for a Tor transport for libp2p.
+#[derive(Clone)]
+pub struct TorTcpConfig {
+    inner: GenTcpConfig<Tcp>,
+    /// Tor SOCKS5 proxy port number.
+    socks_port: u16,
+}
+
+impl TorTcpConfig {
+    pub fn new(tcp: TokioTcpConfig, socks_port: u16) -> Self {
+        Self {
+            inner: tcp,
+            socks_port,
+        }
+    }
+}
+
+impl Transport for TorTcpConfig {
+    type Output = TcpStream;
+    type Error = io::Error;
+    type Listener = TcpListenStream<Tcp>;
+    type ListenerUpgrade = Ready<Result<Self::Output, Self::Error>>;
+    #[allow(clippy::type_complexity)]
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        self.inner.listen_on(addr)
+    }
+
+    // dials via Tor's socks5 proxy if configured and if the provided address is an
+    // onion address. or it falls back to Tcp dialling
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        async fn do_tor_dial(socks_port: u16, dest: String) -> Result<TcpStream, io::Error> {
+            tracing::trace!("Connecting through Tor proxy to address: {}", dest);
+            let stream = connect_to_socks_proxy(dest, socks_port)
+                .await
+                .map_err(|e| io::Error::new(io::ErrorKind::ConnectionRefused, e))?;
+            tracing::trace!("Connection through Tor established");
+            Ok(stream)
+        }
+
+        match to_address_string(addr.clone()) {
+            Some(tor_address_string) => {
+                Ok(Box::pin(do_tor_dial(self.socks_port, tor_address_string)))
+            }
+            _ => {
+                tracing::warn!(
+                    "Address {} could not be formatted. Dialling via clear net",
```

Comment:
```suggestion
                    "Address {} could not be formatted as onion address. Dialling via clear net",
```

This message might be somewhat confusing for a user. Not sure if it should be on `warn` level - but given that someone might want to stay anonymous it is better to warn them when going clearnet in case they dial `dnsaddr`

Reply:
Regarding your change: `to_address_string` is not formatting it to an onion address but formats it to a string of the format `address:port` so that we can dial through Tor. 

I agree that this might not be confusing to some and was deciding between failing, i.e. not dialing at all, or just falling back to clear net. 
I opted for dialing via clear net in the case we cannot format that address string.